mlos_bench/mlos_bench/tests/environments/composite_env_test.py

```diff
@@ -70,11 +112,38 @@ def test_composite_env_params(composite_env: CompositeEnv) -> None:
         "vmName": "Mock VM",        # const_args from the parent
         "EnvId": 1,                 # const_args from the child
         "vmSize": "Standard_B4ms",  # tunable_params from the parent
+        # "someConst": "root"        # not required, so not passed from the parent
     }
     assert composite_env.children[1].parameters == {
         "vmName": "Mock VM",        # const_args from the parent
         "EnvId": 2,                 # const_args from the child
         "idle": "halt",             # tunable_params from the parent
+        # "someConst": "root"        # not required, so not passed from the parent
+    }
+    assert isinstance(composite_env.children[2], CompositeEnv)
+    # CompositeEnv child should receive everything from the parent since it didn't specify a tunable_params subgroup to filter on.
+    # It should also override the const_args.
+    new_params = composite_env.parameters.copy()
+    new_params["vmName"] = "Nested Mock VM"
```

Comment:
@motus a topic to discuss: right now the way that CompositeEnv and merge_parameters() works, it appears that the parent value overrides the value in the children.
I guess that makes sense for the "placeholder" values.
But for child CompositeEnvs, where we might want to have one side be a client and another side be the server, we need to override it in the child CompositeEnv, and then let that propogate down.
Thoughts?

Reply:
@bpkroth you're right, parameters propagate top to bottom, and in this example *all* `vmName` values in *all* environments regardless of their position in the tree will be set to the same value of `"Mock VM"` as it is specified at the top level. We do it this way so we can pass the global parameters (e.g., `accessToken` or `storageFileShareName`) downstream to leaf environments and services. In our MySQL setup we don't have any conflicts between the top and bottom levels - at the top we provision two VMs with the names parameterized as `vmName_server` and `vmName_client`, and at the next level we use `vmName` (this parameter does not occur anywhere above) to which we assign a copy of either `vmName_server` or `vmName_client` value. Yes, there is duplication, and later we might want to introduce some copy rules, so we can assign, e.g., `"vmName": "$vmName_server"` but for the time being it allows us to provision two VMs and run our experiments. Let's discuss it next week when we have time